zulipterminal/helper.py

```diff
@@ -459,15 +459,48 @@ def match_emoji(emoji: str, text: str) -> bool:
     return emoji.lower().startswith(text.lower())
 
 
-def match_stream(stream: Any, text: str) -> bool:
+def match_stream(streams: Dict[Any, List[str]], text: str,
+                 pinned_streams: List[str]) -> List[str]:
     """
-    True if the stream matches with `text` (case insensitive),
-    False otherwise.
+    Returns a list of streams whose words match with the 'text' in the
+    following order:
+    * 1st-word startswith match > 2nd-word startswith match > ... (pinned)
+    * 1st-word startswith match > 2nd-word startswith match > ... (unpinned)
+
+    Note: This function expects `streams` to be sorted in a non-decreasing
+    order.
     """
-    stream_name = stream[0].lower()
-    if stream_name.startswith(text.lower()):
-        return True
-    return False
+    pinned_stream_names = [stream[0] for stream in pinned_streams]
+
+    delimiters = [' ', '-', '_', '/']
+    for stream in streams:
+        stream_name = streams[stream][0]
+        for delimiter in delimiters:
+            if delimiter in stream_name:
+                streams[stream].extend(stream_name.split(delimiter)[1:])
+                break
+
+    matches = OrderedDict([
+        ('pinned', defaultdict(list)),
+        ('unpinned', defaultdict(list)),
+    ])  # type: Dict[str, Dict[int, List[str]]]
+    for stream in streams:
+        for match_position, word in enumerate(streams[stream]):
+            if word.lower().startswith(text.lower()):
+                kind = 'pinned' if streams[stream][0] in pinned_stream_names \
+                                else 'unpinned'
+                matches[kind][match_position].append(stream)
+
+    ordered_matches = []
+    for kind in matches:
+        if not matches[kind]:
+            continue
+        for match_position in range(max(dict(matches[kind])) + 1):
+            if match_position in matches[kind]:
+                for stream in matches[kind][match_position]:
+                    if stream not in ordered_matches:
+                        ordered_matches.append(stream)
+    return ordered_matches
```

Comment:
Is this a `List[str]`, or `List[Any]`, where it matches the `Any` passed in? If so, you could use a `TypeVar`, both for type-checking and making this clearer.

Reply:
My bad. It is `List[Any]`.

Are you suggesting using something like the following?
```python
T = TypeVar('T', Any, str)
```

Reply:
You can constrain the types in that way, though the one you mention is not very useful - it's saying that `T` can be consistently `Any` or `str`. You may be thinking of `AnyStr` as an example of this.

Really I was indicating how you can relate the function parameter types and return type(s) - if they *are* supposed to be related - by using a `TypeVar`.

The mypy pages describe this quite well I think, and there are examples in zulip/zulip, but I was querying whether this kind of approach was applicable:
```
X = TypeVar('X')
def foo(x: X) -> X:
  ...

foo(5)  # must return int
foo(2.5)  # must return float
foo("y")  # must return str
```

Reply:
Thanks for the explanation. :)